Snake Food.

Stan-Kudri opened the pull request.

Stan-Kudri committed:
Add folder "Components" and moved the classes into it "Field" and "Snake".

Stan-Kudri committed:
Add and Changed:

-Add class Food.
-Snake :
  1)Add const char SymbolEat;
  2)Changed method Intersect;
  3)Add method NotEmptyPoit,  which checks if the cell is empty.
-Point: Add method CoordinateMatch that checks the similarity of coordinates
-Main:
  1)Add method ConsoleViewSetting;
  2)Changed the method Game with the addition of a new class and changes.

blowin commented on Snake/Components/Food.cs:
```diff
@@ -0,0 +1,26 @@
+            _widthField = field.Width;
+        }
+
+        public void Create() {
```
лучше метод назвать Draw и передавать координаты

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -27,18 +28,28 @@ public class Snake {
         }
 
-        public bool Move() {
+        public bool Move(ref Food? food) {
```
Лучше делать какой-то компонент, которы будет отвечать за проверки столкновений, ну и код становиться тут уж очень тяжелый для понимания

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -100,5 +111,15 @@ public class Snake {
             return false;
         }
+
+        private bool NotEmptyPoint(Point food) {
```
не понял смысл названия метода и того, что делает

blowin commented on Snake/Point.cs:
```diff
@@ -26,6 +26,10 @@ public class Point {
             Console.WriteLine(symbol);
         }
 
+        public bool CoordinateMatch(Point point) {
```
Equals?

Stan-Kudri committed:
Changed:

-Class "Food": Rename method from "Creat" to "Draw".
-Class "Snake":
  1)Changed method "NotEmptyPoint" and Rename to "PointBusyItBody";
  2)Delite "food.Eat.Draw()", because it is used in the method "Create" (class Food).
-Clas "Direct": Add constructor without input data.
-Clas "Point": Rename method from "CoordinateMatch" to "EqualsCoordinate";
-Small changed in main.

Stan-Kudri committed:
Changed the input parameter type from nullable and removed the unnecessary condition.

Stan-Kudri committed:
Changed:

-Food: Add method Clear for a remove point from field.
-Snake:
  1)Made the method public "Intersect".
  2Changed the "Move" method, which does not contain a condition "Intersect".
-Program (Main): Changed the "Game" method Changed the method based on the new changes.

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -38,12 +37,11 @@ public class Snake {
             _body.Add(_head);
 
-            if (food != null && food.Eat.CoordinateMatch(_head)) {
+            if (food != null && food.Eat.EqualsCoordinate(_head)) {
```
есть стандартный метод Equals

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -112,14 +110,6 @@ public class Snake {
-
-            return false;
-        }
+        private bool PointBusyItBody(Point food) => _body.Any(x => food.EqualsCoordinate(x));
```
У тебя уже есть похожий метод и ты назвал его Intersect, так почему бы не делать единообразно?

blowin commented on Snake/Direct.cs:
```diff
@@ -4,6 +4,10 @@ namespace GameSnake {
     internal class Direct {
         private Directions _currentDirection;
 
+        public Direct() : this(Directions.Left) {
```
лучге явно передать

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -27,13 +27,23 @@ public class Snake {
         }
 
-        public bool Move(ref Food food) {
+        public bool Intersect() {
```
HeadIntersectWithBody

Stan-Kudri committed:
Smashed method "Move" into two methods "Move" and "EatFood", and changed the method "Game" with their implementation.

Stan-Kudri committed:
Changed:

-Point:
    1)Removed the field _symbol and all actions with it in all to class;
    2)Add an input parameter "(char symbol)" to the "Draw" method.
-Field: Add "const char SymbolFeild" which is transmitted to the method "Draw".
-Food: Changed method "Draw" with new changes "Point".
-Snake: Changed method "Move", "EatFood" and other with new changes "Point".

Stan-Kudri committed:
Changed:

-Food:
  1)Rename property from "Eat" ro "Value";
  2)Add method "New()" which creates a new point;
  3)Changed method "Draw()" - he only draw.
-Snake: Small changed.
-Main:
  1)Rendered methods "ConsoleViewSetting" and "WriteMessage" with method "Game";
  2)Removed the field "gameOver" and implemented the method "Game" without it.

Stan-Kudri committed:
Changed:

Snake:
  1)Make the method "IsFreePoint" public;
  2)Removed the loop from the method "EatFood".
Main: Add a loop removed from the method "EatFood".

Stan-Kudri committed:
The food point is passed to the "EatFood" method.

(ref Food food) -> (Point food)

Stan-Kudri committed:
Moved the loop into a separate method "Loop".

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -43,22 +43,30 @@ public class Snake {
             _body.Add(_head);
+            _body.Remove(tail);
+            tail.Clear();
+            _head.Draw();
```
почему тут Draw?

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -43,22 +43,30 @@ public class Snake {
+        }
 
-            if (food.Eat.EqualsCoordinate(_head)) {
+        public bool EatFood(ref Food food) {
```
почему тут ref если ты не меняешь переменную?

blowin replied:
не помнимаю, почему ты везде вызываешь Draw

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -105,6 +113,6 @@ public class Snake {
         }
 
-        private bool PointBusyItBody(Point food) => _body.Any(x => food.EqualsCoordinate(x));
+        private bool IsFreePoint(Point food) => _body.Any(x => food.EqualsCoordinate(x));
```
почему так назван?

blowin commented on Snake/Program.cs:
```diff
@@ -35,12 +35,12 @@
         }
-        else {
-            snake.Move(ref food);
+        else if (!snake.EatFood(ref food)) {
```
где ковычки?

blowin commented on Snake/Components/Food.cs:
```diff
@@ -18,9 +18,8 @@ public class Food {
-                SymbolEat);
-            Eat.Draw();
+                _random.Next(MinCoordinatePoint, _heightField - 1));
+            Eat.Draw(SymbolEat);
```
почему не сделать объект Food и он будет хранить символ?

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -48,16 +48,16 @@ public class Snake {
             _body.Remove(tail);
             tail.Clear();
-            _head.Draw();
+            _head.Draw(SymbolSnake);
```
почему не сделать компонент Head?

blowin commented on Snake/Program.cs:
```diff
@@ -19,7 +21,7 @@
 
     food.Draw();
-    food.Eat.Draw();
+    food.Eat.Draw(SymbolEat);
```
везде по коду теперь константы, делать отдельные объекты для этих элементов

blowin commented on Snake/Components/Food.cs:
```diff
@@ -6,22 +6,24 @@ public class Food {
         private int _heightField;
         private int _widthField;
 
-        public Point Eat { get; set; }
```
Почему Value, если это координаты?

blowin commented on Snake/Components/Food.cs:
```diff
@@ -6,22 +6,24 @@ public class Food {
 
-        public void Draw() {
-            Eat = new Point(
+        public void New() {
```
Метод плохой, лучше принимай в конструкторе Point и сделай статический метод, который может сгенерировать точку или ещё один конструктор, который вызовет главный и передаст рандомную точку

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -45,22 +45,24 @@ public class Snake {
 
         public bool EatFood(ref Food food) {
-            var foodPoint = food.Eat;
+            var foodPoint = food.Value;
```
тут foodPoint, в еде это какое-то Value

blowin commented on Snake/Components/Snake.cs:
```diff
@@ -45,22 +45,24 @@ public class Snake {
-                while (IsFreePoint(food.Eat)) {
-                    food.Draw();
+                while (IsFreePoint(food.Value)) {
+                    food.New();
```
метод New вообще не говорит о том, что там генерируется точка

blowin commented on Snake/Direct.cs:
```diff
@@ -2,9 +2,9 @@
-        private Directions _currentDirection;
+        private Directions _currentDirection = Directions.Left;
 
-        public Direct() : this(Directions.Left) {
```
почему нельзя самом передать Directions? сделай 1 конструктор
Direct(Directions direction = Directions.Left) - всё

blowin commented on Snake/Direct.cs:
```diff
@@ -2,9 +2,9 @@
 
 namespace GameSnake {
     internal class Direct {
```
Direction

blowin commented on Snake/Program.cs:
```diff
@@ -2,28 +2,24 @@
 var height = 20;
 var width = 40;
 
+ConsoleViewSetting(width, height);
```
DisplaySettings

blowin commented on Snake/Program.cs:
```diff
@@ -2,28 +2,24 @@
 
+ConsoleViewSetting(width, height);
 Game(height, width);
+WriteMessage(width, height); //Massage "Game Over" 
```
DisplayGameOver?

blowin commented on Snake/Program.cs:
```diff
@@ -18,7 +18,26 @@
+    Console.Title = "SNAKE";
+}
 
+static void Loop(Snake snake, Food food, Direct directory) {
```
GameLoop

Stan-Kudri committed:
Changed and Add:

-Add extension class PointExtension (Method "Add" and "Clear").

-Rename class from Direct to Direction.

-Point : Removed method "Draw" and "Clear".

-Main:
  1)Reaname method from "WriteMessage" to "DisplayGameOver";
  1)Reaname method from "ConsoleViewSetting" to "DisplaySettings".

-Snake: Rename method from "IsFreePoint" to "IntersectBody".

-Food: Rename property from "Value" to "Point".

Add Extension class "PointExtension" to classes "Field", "Food" and "Snake".

Stan-Kudri committed:
Changed:

-Add Extension method "NewPosition".
-Direction: Direction.Left -> Direction.Right.
-Snake:
  1)Delite field Direction "_directory", because it is auto property "Directoin";
  2)Add methods "DrawHead" and "ClearTail";
  3)Changed method input parameters 'IntersectBody";
  4)Removed method calls "Draw" and "Clear" in methods "Move" and "EatFood".
-Food: Changed the method for new changes Extension class.
-Main: Using Methods class Snake "ClearTail" and "DrawHead".

Stan-Kudri committed:
Rename method from "Loop" to "GameLoop".

blowin commented on Snake/Extension/FoodExtension.cs:
```diff
@@ -0,0 +1,14 @@
+        private const int MinCoordinatePoint = 1;//Because the border value is 0.
+        private static Random random = new Random();
+
+        public static Point NewPosition(this Field field) {
```
Это GeneratePosition

blowin commented on Snake/Components/Food.cs:
```diff
@@ -3,29 +3,20 @@
-        public Point Point { get; set; }
-
-        private Random _random = new Random();
+        public Point Position { get; set; }
```
по всему коду тогда используй Position

Stan-Kudri committed:
Renamed a variable Point "point" to "position".

Stan-Kudri committed:
Rename property Directions from "Value" to "Turn" and Add/Delite empty line.
